File: LeetCode_120/Main.py
class Solution(object):
    def dfs(self, i, j, triangle, dp, n, inf):
        if i + 1 == n:  
            return triangle[i][j]
        
        if dp[i][j] != inf:
            return dp[i][j]
        
        choice1 = self.dfs(i + 1, j, triangle, dp, n, inf)
        choice2 = self.dfs(i + 1, j + 1, triangle, dp, n, inf)
    
        dp[i][j] = triangle[i][j] + min(choice1, choice2)
        return dp[i][j]

    def minimumTotal(self, triangle):
        """
        :type triangle: List[List[int]]
        :rtype: int
        """
        inf = 10**9
        n = len(triangle)
        dp = [[inf for _ in range(n)] for _ in range(n)]

        return self.dfs(0, 0, triangle, dp, n, inf)

File: LeetCode_120/Main.js
/**
 * @param {number[][]} triangle
 * @return {number}
 */

const dfs = function(i, j, triangle, dp, n, inf) {
    if (i === n - 1) {
        return triangle[i][j]; 
    }
    if (dp[i][j] !== inf) {
        return dp[i][j];
    }
    const choice1 = dfs(i + 1, j, triangle, dp, n, inf);
    const choice2 = dfs(i + 1, j + 1, triangle, dp, n, inf);

    return dp[i][j] = triangle[i][j] + Math.min(choice1, choice2);
}

var minimumTotal = function(triangle) {
    const n = triangle.length;
    const inf = 1e9;
    
    const dp = Array.from({ length: n }, () => Array(n).fill(inf));

    return dfs(0, 0, triangle, dp, n, inf);
};

console.log(minimumTotal([[2],[3,4],[6,5,7],[4,1,8,3]]));
console.log(minimumTotal([[-10]]));
